src/ck-core/cksyncbarrier.C

```diff
@@ -131,39 +145,33 @@ void CkSyncBarrier::propagate_atsync()
 
 void CkSyncBarrier::recvLbStart(int lb_step, int sourcenode, int pe)
 {
-  if (lb_step != cur_epoch || startedAtSync) return;
+  if (lb_step > currentKick)
+    currentKick = lb_step;
+  if (lb_step < cur_epoch || startedAtSync)
+    return;
+
   const int mype = CkMyPe();
   const int mynode = CkNodeOf(mype);
   if (sourcenode < mynode)
     received_from_left = true;
   else if (sourcenode > mynode)
     received_from_right = true;
   else if (rank0pe) // convert incoming pe number to local rank
-    rank_needs_flood[pe - mype] = false;
+    rank_needs_kick[pe - mype] = false;
   else
     received_from_rank0 = true;
-  if (clients.size() == 1 &&
-      clients.front()->chare->isLocMgr())  // CkLocMgr is usually a client on each PE
-    CheckBarrier(true);  // Empty PE invokes barrier on self on receiving a flood msg
+
+  if (clients.empty())
+    CheckBarrier();  // Empty PE invokes barrier on self on receiving a kick
 }
 
-void CkSyncBarrier::CheckBarrier(bool flood_atsync)
+void CkSyncBarrier::CheckBarrier()
 {
   if (!on) return;
 
   const auto client_count = clients.size();
 
-  if (client_count == 1 && !flood_atsync)
-  {
-    if (clients.front()->chare->isLocMgr()) return;
-  }
-
-  // If there are no clients, resume as soon as we're turned on
-  if (client_count == 0)
-  {
-    cur_epoch++;
-    CallReceivers();
```

Comment:
This CallReceivers() calls ResumeFromSync on chares for Periodic LB I think. It probably needs to be added back to the code. Also, with LB refactor I think the part which corresponds to this, that added receivers for Periodic LB (!usesAtSync) was incorrectly removed from ckFinishConstruction() in cklocation.C.

Reply:
Should `ResumeFromSync` be called in periodic LB? The documentation only states that it is called when using `AtSync` mode, which I think is the only sensible mode to call it in, since it's the only one that offers a causality chain for it.

Reply:
Sounds good, I guess we can remove CallReceivers() if there's no use case for it. I am not sure what the previous use case was, but the older code before LB refactor used to call ResumeFromSync (if implemented) for periodic LB.

Reply:
Okay, I think removing it makes use cleaner. I think the previous behavior was incidental and not an intentional decision.